pageserver/src/walredo.rs

```diff
@@ -90,6 +92,21 @@ pub trait WalRedoManager: Send + Sync {
     ) -> Result<Bytes, WalRedoError>;
 }
 
+struct ProcessInput {
+    child: NoLeakChild,
+    stream: ChildStdin,
+    stdout_fd: RawFd,
+    stderr_fd: RawFd,
+    stdin_fd: RawFd,
```

Comment:
I wonder if this could be clarified a bit by having only one field for `stdin` instead of generic `stream`. Using the `as_raw_fd()` should be as fast as reading it from this field.

Another point of clarification could be the removal of the `stdout_fd` since it's never polled or read while having access to `ProcessInput`.

```suggestion
    /// Polled and written to while holding access to `ProcessInput`.
    stdin: ChildStdin,
    /// Polled and read from by request writing and response reading parts.
    ///
    /// An owned version is not needed, because the pipe will be closed by `NoLeakChild::drop`.
    stderr_fd: RawFd,
```

This would leave us with only one fd shared between writing to stdin or waiting from stdout. Similarly the `stderr_fd` could be added to `ProcessOutput`.

This would open up refactoring possibilities for example splitting the code calling `poll` into methods on `ProcessInput` and `ProcessOutput`, and eventually deduplicating the stderr handling, for example:

```rust
impl ProcessInput {
  fn poll(&self, remaining: Duration) -> std::io::Result<(StdinStatus, StderrStatus)> {
    loop {
      ...
    }
    // return timeout error from here
  }
}

impl ProcessOutput {
  fn poll(&self, remaining: Duration) -> std::io::Result<(StdoutStatus, StderrStatus)> {
    loop {
      ...
    }
    // return timeout error from here
  }
}

struct StdinStatus(RawFd, PollFlags);

impl StdinStatus {
  fn continue_write(&self, buf: &[u8]) -> Option<std::io::Result<usize>> {
    // if not writeable or errored, return None
    // otherwise try write again, return Some(result)
    //
    // alternatively this could just impl std::io::Write but check
    // revents before `write`, returning `Err(WouldBlock)` which would have 
    // to be ignored similar to the None variant.
  }
}

struct StdoutStatus(RawFd, PollFlags);

impl StdoutStatus {
  fn continue_read(&self, buf: &mut [u8]) -> Option<std::io::Result<usize>> {
    // implemented similarly to StdinStatus::continue_write
  }
}

struct StderrStatus(RawFd, PollFlags);

impl StdErrStatus {
  fn log_any_output(&self, err: &Mutex<Option<ChildStderr>>) -> std::io::Result<()> {
    let revents = self.1.revents().expect("unknown flags returned for stderr");
    if revents & (PollFlags::POLLERR | PollFlags::POLLIN) == PollFlags::empty() {
      return Ok(());
    }

    // could use File::from_raw_fd and std::mem::forget(file) to avoid using a different fd from the polled one
    // in case it has been restarted already
    let err = err.lock().unwrap().as_mut().expect("ChildStderr should had been set");

    // the currently duplicated reading and logging
  }
}
```

Perhaps apart from the race to `Mutex<Option<ChildStd(out|in|err)>>` this could all be a follow-up. I will make a different note about the race.

Reply:
Partly done,
I prefer to postpone this refactoring and did them in another PR.
We in any case may need to replace synchronous  `nix::poll::poll` with some Tokio analog.